spec/weather_retriever_spec.rb

```diff
@@ -0,0 +1,224 @@
+# frozen_string_literal: true
+
+require 'open_weather_client'
+require 'city_name_converter'
+require 'sql_database'
+require 'weather_retriever'
+
+RSpec.describe WeatherRetriever do
+  let(:presenter) { double('presenter') }
+  let(:sql) { SQLite3::Database.open('file::memory:?cache=shared') }
+  let(:today) { Time.utc(2021, 4, 9, 13, 0, 1) }
+  let(:weather_info) { File.open('fixtures/london_weather.json').read }
+
+  before do
+    allow(Time).to receive(:now).and_return today
+    presenter_stub
+  end
+
+  after { sql.execute 'DROP TABLE weather' }
+
+  describe '#run' do
+    it 'collects the city name' do
+      open_weather_stub
+      set_up_retriever.run
+
+      expect(presenter).to have_received(:city_name).once
+    end
+
+    context 'When successful' do
+      before { open_weather_stub }
+
+      context 'when database is up to data' do
+        let(:earier_today) { Time.utc(today.year, today.month, today.day, today.hour - 3).to_i }
+
+        before do
+          create_table
+          setup_name_index
+          add_weather_row(earier_today)
+        end
+
+        it 'does not call the client' do
+          set_up_retriever.run
+
+          expect(open_weather_stub).not_to have_been_made
+        end
+
+        it 'does not augment the database' do
+          set_up_retriever.run
+
+          result = sql.query "SELECT * FROM weather WHERE name = 'London'" do |rows|
+            rows.next_hash.transform_keys(&:to_sym)
+          end
+
+          expect(result[:unix_date]).to eq(earier_today)
+        end
```

Comment:
I've just tested this to be sure. when the database augments data with the same city name it replaces it, so it will still have 1 row. That is why I am directly testing the data in the row.